nsiqcppstyle_types.py

```diff
@@ -0,0 +1,69 @@
+# Copyright (c) 2022 All rights reserved.
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#    * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#    * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from typing import NewType, Union
+import nsiqcppstyle_lexer
+import nsiqcppstyle_checker
+
+# A stack showing the lexical context the token resides in
+ContextStackType = NewType('ContextStackType', nsiqcppstyle_checker.ContextStack)
+
+# The new context entry that will be pushed to the contextStack due to this token
+ContextType = NewType('ContextType', nsiqcppstyle_checker.Context)
+
+# A boolean flag denoting whether the function is a declaration (True) or definition (False)
+DeclarationType = NewType('DeclarationType', bool)
+
+# A directory name
+DirNameType = NewType('DirNameType', str)
+
+# The name of the file being analyzed
+FileNameType = NewType('FileNameType', str)
+
+# A full function name
+FullNameType = NewType('FullNameType', str)
```

Comment:
I'd prefer these to be valid Path to remove a bunch of corner cases due to using str (eg: some filesystems don't allow unicode names, but others do), but let's postpone that to later

Reply:
If we type it here as "Path", would we need to modify the engine code to define all file paths as Path objects, for consistency reasons?

Reply:
I don't think the engine internals need to be updated. Just the usage boundaries. This could be a slow work over several PRs